### PR TITLE
[CI] Build wheels and publish to pypi

### PR DESCRIPTION
**Pull request by Skydes**
At each PR or push to master, wheels for Linux and macOS are built with GitHub Actions and available [here](https://github.com/mihaidusmanu/pycolmap/actions). Optionally, we automatically publish them to pypi. Build time is currently ~25min.

Based on @johnwlambert excellent repo https://github.com/johnwlambert/pycolmap-wheels with some differences:
- fix some issues with wheel filenames
- pass Eigen3 (for CentOS) and Qt5 (for macOS) flags to `setup.py`
- add package version as module attribute
- use C++14 when Ceres>=2.0
- automatically publish to pypi if a new tag is pushed (e.g. by a new release)

TODO:
- [x] cleanup the *.sh scripts
- [x] make macOS work reliably

Any help is very much appreciated.

## Note on macOS

I am on 11.4 but `python3 -c "from distutils import util; print(util.get_platform())"` returns `macosx-10.15-x86_64`. Building with the target `macos-11` produces `pycolmap-0.0.2-cp39-cp39-macosx_11_0_x86_64.whl`. This is rejected by pip unless I rename it to `pycolmap-0.0.2-cp39-none-macosx_10_9_x86_64.whl`. Then the macOS wheel work perfectly. I have no idea why this happens and why the abi tag needs to be `none`.

cc @mihaidusmanu @Phil26AT @johnwlambert

### CODE REVIEW

File: .github/workflows/build-new.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      run: |
+        ./package/build-wheels-macos.sh ${{ matrix.pyversion }}
+    - name: Archive wheels
+      uses: actions/upload-artifact@v2
+      with:
+        # we strip the version number from the artifact name
+        name: pycolmap-${{ matrix.pyversion }}-macosx_11_0_x86_64
+        path: ./wheelhouse/pycolmap-*-macosx_11_0_x86_64.whl
+    - name: Publish package
+      if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
```
Comment (johnwlambert) [inline .github/workflows/build-new.yml:70]: Hi @Skydes thanks for the PR. Very cool. Could you maybe add an inline comment here explaining about under which criteria the package is automatically published?

File: package/build-wheels-linux.sh [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+fi
+set -e -x
+
+make -j$(nproc) install
+
+mkdir -p /io/wheelhouse
+
+# ----------- Build pycolmap wheel -----------------------------------------------------
+cd /io/
+#cp package/setup_centos.py setup.py
```
Comment (johnwlambert) [inline package/build-wheels-linux.sh:211]: Very nice fix in the setup.py. Can we remove this commented out code now?

File: package/build-wheels-linux.sh [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+cat setup.py
+
+PLAT=manylinux2014_x86_64
+#"${PYBIN}/python" setup.py bdist_wheel --python-tag=$PYTHONVER --plat-name=$PLAT
+EIGEN3_INCLUDE_DIRS="/eigen-3.3.9" "${PYBIN}/python" setup.py bdist_wheel --plat-name=$PLAT #--python-tag=$PYTHONVER
+
+# Bundle external shared libraries into the wheels
+for whl in ./dist/*.whl; do
+    auditwheel repair "$whl" -w /io/wheelhouse/ --no-update-tags
```
Comment (johnwlambert) [inline package/build-wheels-linux.sh:220]: Nice -- curious what `--no-update-tags` does?

File: setup.py [hunk 4]
```diff
@@ -33,8 +33,17 @@ def run(self):
...
-                      '-DPYTHON_EXECUTABLE=' + sys.executable]
+        cmake_args = [
+            '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=' + extdir,
+            '-DPYTHON_EXECUTABLE=' + sys.executable,
+            '-DVERSION_INFO={}'.format(self.distribution.get_version()),
+        ]
+        eigen_dir = os.environ.get('EIGEN3_INCLUDE_DIRS')
+        if eigen_dir is not None:
+            cmake_args += ['-DEIGEN3_INCLUDE_DIRS={}'.format(eigen_dir)]
+        qt5_dir = os.environ.get("Qt5_DIR")
```
Comment (johnwlambert) [inline setup.py:44]: beautiful! glad to see my old hacks gone

File: .github/workflows/build-new.yml [hunk 5]
```diff
@@ -0,0 +1,75 @@
+name: Pybind11 CI
+
+on:
+  push:
+    branches:
+      - master
+      - wheels
```
Comment (johnwlambert) [inline .github/workflows/build-new.yml:7]: just curious about the plan for the wheels branch vs. master branch.

Comment (Skydes) [reply .github/workflows/build-new.yml]: I just wanted to trigger the CI before opening the PR. Unnecessary now.

File: package/build-wheels-linux.sh (hunk 3 above)
Comment (Skydes) [reply package/build-wheels-linux.sh:188]: Without `--no-update-tags`, auditwheel updates the wheel filename by appending alias tags (e.g. `manylinux_2_17_x86_64` for `manylinux2014_x86_64`) so we end up for example with `pycolmap-0.0.2-cp37-cp37m-manylinux2014_x86_64.manylinux_2_17_x86_64.whl`. This is confusing and can break pip/pypi checks.

File: package/build-wheels-macos.sh [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+for compiler in cc c++ gcc g++ clang clang++
+do
+    which $compiler
+    $compiler --version
+done
+
+echo "Python versions: ${PYTHON_VERS[@]}"
+# Compile wheels
+for PYVER in ${PYTHON_VERS[@]}; do
```
Comment (Skydes) [inline package/build-wheels-macos.sh:80]: @johnwlambert Why do we need to loop over versions here? The CI calls this script for each different value of `pyversion` so `PYTHON_VERS` is always a single version string, not a list.

File: .github/workflows/build-new.yml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    types: [ assigned, opened, synchronize, reopened ]
+  release:
+    types: [published]
+  workflow_dispatch:
+
```
Comment (Skydes) [inline .github/workflows/build-new.yml:12]: @johnwlambert Your repo has an additional trigger `repository_dispatch:    types: [pybind-wrapper]` here, not sure what this means so I omitted it.

File: package/build-wheels-linux.sh (hunk 3 above)
Comment (johnwlambert) [reply package/build-wheels-linux.sh:188]: Got it -- thanks!

File: package/build-wheels-macos.sh (hunk 6 above)
Comment (johnwlambert) [reply package/build-wheels-macos.sh:80]: good catch -- yes the loop is a no-op and can be removed

**Comment by Skydes** [comment]
Update: apparently installing the 10.15 wheel on my macOS 11.4 now works, and I guess that it will pull the right one on pypi. I'll merge once I have permissions to add the pypi token to the repo @mihaidusmanu

**Review by johnwlambert — APPROVED** [review]
LGTM, feel free to merge when you're ready. Thanks Paul-Edouard.

On macOS 10.15, with the wheel `pycolmap-0.0.2-cp38-cp38-macosx_10_15_x86_64.whl`, when I run it, I see:
```bash
OMP: Error #15: Initializing libomp.dylib, but found libiomp5.dylib already initialized.
OMP: Hint This means that multiple copies of the OpenMP runtime have been linked into the program. That is dangerous, since it can degrade performance or cause incorrect results. The best thing to do is to ensure that only a single OpenMP runtime is linked into the process, e.g. by avoiding static linking of the OpenMP runtime in any library. As an unsafe, unsupported, undocumented workaround you can set the environment variable KMP_DUPLICATE_LIB_OK=TRUE to allow the program to continue to execute, but that may cause crashes or silently produce incorrect results. For more information, please see http://openmp.llvm.org/
zsh: abort      ...
```
(I saw this before on the macOS wheels I built previously). But when I set the following environment variable, things proceed and work just fine.
```bash
export KMP_DUPLICATE_LIB_OK=TRUE
```
Not a blocker now, but something for us to look into long-term.

**Pull request merged**